documentation/specs/static-graph.md

```diff
@@ -227,34 +141,109 @@ We'll represent the project reference protocols as items in MSBuild. The target
+
+        // Projects which this project references.
+        IReadOnlyCollection<ProjectGraphNode> ProjectReferences { get; }
+
+        // Projects which reference this project.
+        IReadOnlyCollection<ProjectGraphNode> ReferencingProjects { get; }
+
+        // The evaluated project
+        Project Project { get; }
+    }
+}
+```
 
-**Open Issue:** With this approach initial targets will also get included due to how the RequestBuilder composes entry targets and how TargetBuilder reports build results. Is this an issue? Is it just a perf optimization to take them out from the cache?
+## Isolated builds
+Building a project in isolation means that any build results for project references must be pre-computed and provided as input.
 
-For projects which depend on other projects, the MSBuild task will look up the target result in the cache. If the target is missing from the cache, an error will be logged and the build will fail. If the project is calling into itself either via `CallTarget` or via `MSBuild` with a different set of global properties, this will be always allowed (support for cross-targeting).
+If a project uses the MSBuild task, the build result must be in MSBuild's build result cache instead of just-in-time executing targets on that referenced project. If it is not in the build result cache, an error will be logged and the build will fail. If the project is calling into itself either via `CallTarget` or the MSBuild task with a different set of global properties, this will be allowed to support multitargeting and other build dimensions implemented in a similar way.
 
 Because referenced projects and their entry targets are guaranteed to be in the cache, they will not build again. Therefore we do not need to set `/p:BuildProjectReferences=false` or any other gesture that tells SDKs to not do recursive operations.
 
-Initially this result cache will just be in-memory, similar to the existing MSBuild target result cache, but eventually could persist across builds so that an incremental build of a particular project in the project graph would not require re-evaluation or any target execution in that project's dependencies. This may have significant performance improvements for incremental builds.
+**Open Issue:** With this approach initial targets will also get included due to how the `RequestBuilder` composes entry targets and how `TargetBuilder` reports build results. Is this an issue? Is it just a perf optimization to take them out from the cache? It seems like target mapping incompleteness at the very least, since if the project changes its initial targets, the cache hit from the referencing project would suddenly be a miss.
+
+**Open Issue:** How do we differentiate between targets called on the outer build project and targets called on the inner build projects? For multitargeting builds, a project will actually reach into the inner build, but the graph expresses a dependency on the outer build.
```

Comment:
It could be as simple as "impose a new restriction on outer builds that they have forwarding targets for each inner build target". But that might be a high burden. Could also potentially have a binary outer/inner distinction in the targets (but I don't know how to express that).

Reply:
What I'm trying to explain here is that because we only cache the entry targets, none of the inner build targets would be available to other projects in isolated mode. So although Build with no global props would be in the cache, Build with a specific target framework would not. Well, it would be in the cache, but the isolated build logic would not allow it.

So it seems like for isolated builds, we're going to have to special-case multitargeting regardless, either by splitting the graph into distinct nodes per TFM or in the isolated build cache lookup case.